### PR TITLE
Accept instances in DIMACS format

### LINKED ISSUE #29: [Testing] Accept instances in DIMACS format
DIMACS is a research institute that, among other things, publishes datasets and solutions to a number of common algorithmic problems. In particular, DIMACS publishes a number of datasets for MIS which would be extremely useful for testing this library.

We'd like the ability to load MIS datasets.

This issue has two goals.

- [ ] Create a function `load_dimacs` (see below);
- [ ] Use this function to write several tests using some datasets from [here](https://oeis.org/A265032/a265032.html).

---

Function `load_dimacs`:

```py
@dataclass
class DIMACSDataset:
   instance: MISINstance
   solutions: list[list[int]]

def load_dimacs(path: str) -> DIMACSDataset:
   ...
```

This function loads a dataset from a path (relative to the root of the project).

Here's [an example](https://oeis.org/A265032/a265032_1dc.64.txt.gz) of a file we'd like to read. Note that the files are provided in gzip format but function `load_dimacs` does not need to implement gzip decompression.

For more details on the format, see https://oeis.org/A265032/a265032.html .

---

# Hackathon instructions

- Please test your code before requesting a review!
- If you have any questions, don't hesitate to reach out [on our slack](https://join.slack.com/t/pasqal-community/shared_invite/zt-2y7drym29-ZEH_uZlasEUDhrG9Tk0Vuw)!

---

# AI Warning

If you wish to use some AI to help you write the code, that's your choice. However, it is our experience that most pure AI contributions simply don't work and waste everybody's time, so **please check that the code you provide works as intended**.

### PR DESCRIPTION
**Pull request by y-richie-y**
Closes #29 
The tests run slowly. Is this expected?

### CODE REVIEW

File: tests/test_dimacs.py [hunk 1]
```diff
@@ -0,0 +1,37 @@
+from mis.data.graphs import load_dimacs
+
+
+def test_hexagon():
```
Comment (RolandMacDoland) [inline tests/test_dimacs.py:4]: I guess you can also check that the solution found is genuinely and independent set by using [this method](https://github.com/pasqal-io/maximum-independent-set/blob/08b0b9ff6875acd7dc344c60e2be5225cc8ed4fa/mis/pipeline/kernelization.py#L58).

**Review by RolandMacDoland — COMMENTED** [review]
Thanks @y-richie-y LGTM. I'll let my colleague @Yoric having the final word on this.

**Comment by Yoric** [comment]
> The tests run slowly. Is this expected?

I haven't noticed any slow test. Which test are you referring to?

File: mis/data/dimacs/a265032_1dc.64.txt
Comment (Yoric) [inline mis/data/dimacs/a265032_1dc.64.txt:1]: This data is solely for tests, right? If so, I'd prefer if it went into some subdirectory of `tests/`.

File: mis/data/graphs.py [hunk 2]
```diff
@@ -538,3 +540,39 @@ def ingest(self, graph: NXWithPos, device: pl.devices.Device, id: int) -> BaseGr
         pyg.pos = torch.tensor(positions, dtype=torch.float)
 
         return BaseGraph(id=id, device=device, data=pyg)
+
```
Comment (Yoric) [inline mis/data/graphs.py:543]: Note that file `graphs.py` was removed by another PR, but you can recreate it.

File: mis/data/graphs.py [hunk 3]
```diff
@@ -538,3 +540,39 @@ def ingest(self, graph: NXWithPos, device: pl.devices.Device, id: int) -> BaseGr
...
+
+@dataclass
+class DIMACSDataset:
```
Comment (Yoric) [inline mis/data/graphs.py:546]: Let's take the opportunity to add some documentation here.

File: mis/data/graphs.py [hunk 4]
```diff
@@ -538,3 +540,39 @@ def ingest(self, graph: NXWithPos, device: pl.devices.Device, id: int) -> BaseGr
...
+        DIMACSDataset: An instance of DIMACSDataset.
+    """
+    with open(path, 'r') as f:
+        lines = f.readlines()
+
+    # Parse the graph from the DIMACS format
+    edges = []
+    for line in lines:
+        if line.startswith('p'):
+            continue  # Skip problem line
```
Comment (Yoric) [inline mis/data/graphs.py:568]: @RolandMacDoland Do you think it would be useful to actually load this line and test that we have the right number of edges ad nodes?

File: mis/data/graphs.py [hunk 5]
```diff
@@ -538,3 +540,39 @@ def ingest(self, graph: NXWithPos, device: pl.devices.Device, id: int) -> BaseGr
...
+        if line.startswith('e'):
+            parts = line.split()
+            edges.append((int(parts[1]), int(parts[2])))
+
+    graph = nx.Graph()
+    graph.add_edges_from(edges)
+    instance = MISInstance(graph)
+    solver = MISSolver(instance, SolverConfig())
+    solutions = [e.nodes for e in solver.solve().result()]
```
Comment (Yoric) [inline mis/data/graphs.py:577]: Ah, I realize that the issue description wasn't clear. We don't want to call the solver here. It's just that some DIMACS files contain a solution (or perhaps more than one solution, I haven't checked). `solutions` should be the solution lines.

@RolandMacDoland do you have any such file at hand, for testing?

File: mis/data/graphs.py (hunk 4 above)
Comment (RolandMacDoland) [reply mis/data/graphs.py:568]: Yes, a more elaborated version can be found [here](https://github.com/pasqal-io/maximum-independent-set/pull/28/files#diff-e52e4ddd58b7ef887ab03c04116e676f6280b824ab7469d5d3080e5cba4f2128)

File: mis/data/graphs.py (hunk 5 above)
Comment (RolandMacDoland) [reply mis/data/graphs.py:577]: Do you mean solutions ? If so no. Sometimes DIMACS files contain them that I use in tests. However, there is no guarantee that the current solver returns the same set of node labels.

File: mis/data/dimacs/a265032_1dc.64.txt
Comment (y-richie-y) [reply mis/data/dimacs/a265032_1dc.64.txt:1]: Ok done

File: tests/test_dimacs.py (hunk 1 above)
Comment (y-richie-y) [reply tests/test_dimacs.py:4]: Ok done

File: mis/data/graphs.py (hunk 2 above)
Comment (y-richie-y) [reply mis/data/graphs.py:543]: Ok done. I only added the necessary classes.

File: mis/data/graphs.py (hunk 3 above)
Comment (y-richie-y) [reply mis/data/graphs.py:546]: Ok done

File: mis/data/graphs.py (hunk 4 above)
Comment (y-richie-y) [reply mis/data/graphs.py:568]: Ok done

File: mis/data/graphs.py (hunk 5 above)
Comment (Yoric) [reply mis/data/graphs.py:63]: @RolandMacDoland I meant, do you have an example file that contains the solution? To confirm the syntax.

**Review by RolandMacDoland — COMMENTED** [review]
Thanks @y-richie-y looks good to me. Minor comments so far. Happy to approve once addressed/resolved. Ping @Yoric for another pair of eyes.

**Comment by RolandMacDoland** [comment]
@y-richie-y Could you please rebase ? I've added couple of changes you might want to use. Thanks.

File: mis/data/graphs.py [hunk 6]
```diff
@@ -0,0 +1,64 @@
...
+
+
+@dataclass
+class DIMACSDataset:
+    """
+    A dataset representing a DIMACS graph instance and its solutions.
+    This is used to load DIMACS files and extract the graph and solutions.
+    """
+    instance: MISInstance
+    solutions: list[list[int]]
```
Comment (Yoric) [inline mis/data/graphs.py:23]:
```suggestion
    solutions: list[list[int]]
    """
    If the dataset provided solutions, the list of solutions.
    Note that most instances do not provide a solution and those that do do not guarantee that they provide all solutions.
    """
```

File: mis/data/graphs.py [hunk 7]
```diff
@@ -0,0 +1,64 @@
...
+        if line.startswith('e'):
+            parts = line.split()
+            edges.append((int(parts[1]), int(parts[2])))
+
+    graph = nx.Graph()
+    graph.add_edges_from(edges)
+
+    if graph.number_of_nodes() != n_vertices:
+        raise ValueError(
+            "DIMACS file does not match specified number of vertices."
```
Comment (Yoric) [inline mis/data/graphs.py:54]:
```suggestion
            "Error in DIMACS file: the graph within this file does not have the number of vertices it claims."
```

(slight rephrasing)

File: mis/data/graphs.py [hunk 8]
```diff
@@ -0,0 +1,64 @@
...
+        )
+    if graph.number_of_edges() != n_edges:
+        raise ValueError(
+            "DIMACS file does not match specified number of edges."
```
Comment (Yoric) [inline mis/data/graphs.py:58]:
```suggestion
            "Error in DIMACS file: the graph within this file does not have the number of edges it claims."
```

(slight rephrasing)

File: mis/data/graphs.py (hunk 5 above)
Comment (Yoric) [reply mis/data/graphs.py:63]: @y-richie-y Could you fix this?

File: tests/data/dimacs/square.txt
Comment (Yoric) [inline tests/data/dimacs/square.txt:1]: Why is this file empty? Did you forget to remove it?

**Review by Yoric — CHANGES_REQUESTED** [review]
Almost there :)

File: mis/data/graphs.py (hunk 5 above)
Comment (y-richie-y) [reply mis/data/graphs.py:67]: Sure. What do you want me to do when there is no solution line?

File: tests/test_files/dimacs/graphs.py [hunk 9]
```diff
@@ -0,0 +1,77 @@
...
+    """
+    instance: MISInstance
+    solutions: list[list[int]]
+    """
+    If the dataset provided solutions, the list of solutions.
+    Note that most instances do not provide a solution and those that do do not guarantee that they provide all solutions.
+    """
+
+
+def load_dimacs(path: str) -> DIMACSDataset:
```
Comment (Yoric) [inline tests/test_files/dimacs/graphs.py:31]: Agreed with Roland, using pathlib seems like a good idea.

File: tests/test_files/dimacs/graphs.py
Comment (Yoric) [inline tests/test_files/dimacs/graphs.py:1]: Let's move that file to `mis/data/graphs.py`.

File: tests/test_files/dimacs/graphs.py [hunk 10]
```diff
@@ -0,0 +1,77 @@
...
+    if graph.number_of_edges() != n_edges:
+        raise ValueError(
+            "Error in DIMACS file: the graph within this file does not have the number of edges it claims."
+        )
+
+    instance = MISInstance(graph)
+
+    if solutions is None:
+        solver = MISSolver(instance, SolverConfig())
+        solutions = [e.nodes for e in solver.solve().result()]
```
Comment (Yoric) [inline tests/test_files/dimacs/graphs.py:76]: As just discussed in the PR, let's make it `solutions = []`.

**Review by Yoric — APPROVED** [review]
Good for me with trivial changes.

Let's see if @RolandMacDoland has other ideas :)

**Comment by y-richie-y** [comment]
All done!

File: tests/test_classical_mis.py [hunk 11]
```diff
@@ -107,3 +182,47 @@ def test_star_mis(
...
+    instance = MISInstance(graph)
+    solver = MISSolver(instance, config)
+    solutions = solver.solve().result()
+
+    assert graph.number_of_nodes() == num_nodes
+    assert graph.number_of_edges() == num_edges
+    assert len(solutions) > 0
+    for solution in solutions:
+        assert Kernelization(graph).is_independent(solution.nodes)
+        assert len(solution.nodes) <= max_independent_set_size
```
[the remark this answers is not in the document]
Comment (y-richie-y) [reply tests/test_classical_mis.py:228]: I think they found the optimal solutions already

[the remark this answers is not in the document]
Comment (RolandMacDoland) [reply tests/test_classical_mis.py:228]: OK if you can leave out a comment, that'd be appreciated.

File: tests/test_classical_mis.py [hunk 12]
```diff
@@ -107,3 +182,47 @@ def test_star_mis(
...
         assert 1 <= node < SIZE
+
+
+@pytest.fixture(params=[
+    ("tests/test_files/dimacs/petersen.txt", 10, 15, 5),
+    ("tests/test_files/dimacs/a265032_1dc.64.txt", 64, 543, 10),
+    ("tests/test_files/dimacs/a265032_1tc.32.txt", 32, 68, 12),
+    ("tests/test_files/dimacs/hexagon.txt", 6, 6, 3),
+])
+def graph_data(request):
```
[the remark this answers is not in the document]
Comment (y-richie-y) [reply tests/test_classical_mis.py:193]: not really sure what you mean. I'm verifying the solutions here so I need to additionally pass information about the graph (vertex and edge count, size of solution) to test.

**Review by RolandMacDoland — APPROVED** [review]
Thanks @y-richie-y. LGTM, a final comment about the use of `pathlib` that you can quickly address before merging.

File: mis/data/graphs.py [hunk 13]
```diff
@@ -0,0 +1,64 @@
...
+class DIMACSDataset:
+    """
+    A dataset representing a DIMACS graph instance and its solutions.
+    This is used to load DIMACS files and extract the graph and solutions.
+    """
+    instance: MISInstance
+    solutions: list[list[int]]
+
+
+def load_dimacs(path: str) -> DIMACSDataset:
```
[the remark this answers is not in the document]
Comment (y-richie-y) [reply mis/data/graphs.py:30]: I believe I have addressed this? I'm using `Path` from `pathlib` within the function.

[the remark this answers is not in the document]
Comment (Yoric) [reply mis/data/graphs.py:30]: GitHub still shows me `path: str`.

File: tests/test_classical_mis.py [hunk 14]
```diff
@@ -181,3 +183,44 @@ def test_star_mis(
 
     for node in nodes:
         assert 1 <= node < SIZE
+
+
+@pytest.mark.parametrize(
+    "dimacs_to_nx",
+    [
+        ("tests/test_files/dimacs/petersen.txt", 10, 15, 5),
```
Comment (RolandMacDoland) [inline tests/test_classical_mis.py:191]: Still needs to be harmonised.

**Review by RolandMacDoland — COMMENTED** [review]
Thanks @y-richie-y. If you can address the final comments and resolve pending discussions, you should be good to go.

**Review by RolandMacDoland — APPROVED** [review]
Thanks @y-richie-y. LGTM :) If you want to merge, could you please resolve the discussion threads still open ?

**Comment by y-richie-y** [comment]
Done
